src/lib/storage/dictionary_segment.cpp

```diff
@@ -1,95 +1,157 @@
+  auto unique_values = std::map<T, ValueID>();
+
+  auto last_index = _segment_nullable ? 1 : 0;
+  for (auto index = size_t{0}; index < value_segment_size; ++index) {
+    if (!value_segment->is_null(index)) {
+      auto value = values[index];
+      if (unique_values.find(value) == unique_values.end()) {
+        unique_values[value] = last_index;
+        ++last_index;
+      }
+    }
+  }
+
+  for (auto [value, key] : unique_values) {
+    _dictionary.push_back(value);
+  }
+
+  const auto attribute_vector = get_attribute_vector(unique_values.size(), value_segment_size);
+
+  for (auto index = size_t{0}; index < value_segment_size; ++index) {
+    if (value_segment->is_null(index)) {
+      attribute_vector->set(index, null_value_id());
+    } else {
+      auto dict_value = unique_values[values[index]];
+      attribute_vector->set(index, dict_value);
+    }
+  }
+  _attribute_vector = attribute_vector;
+  return;
 }
 
 template <typename T>
 AllTypeVariant DictionarySegment<T>::operator[](const ChunkOffset chunk_offset) const {
-  // Implementation goes here
-  Fail("Implementation is missing.");
+  auto return_value = get_typed_value(chunk_offset);
+  if (return_value) {
+    return *return_value;
+  }
+  return NULL_VALUE;
 }
 
 template <typename T>
 T DictionarySegment<T>::get(const ChunkOffset chunk_offset) const {
-  // Implementation goes here
-  Fail("Implementation is missing.");
+  const auto return_value = get_typed_value(chunk_offset);
+  Assert(return_value.has_value(), "Value at position: " + std::to_string(chunk_offset) + "is NULL!");
+  return *return_value;
 }
 
 template <typename T>
 std::optional<T> DictionarySegment<T>::get_typed_value(const ChunkOffset chunk_offset) const {
-  // Implementation goes here
-  Fail("Implementation is missing.");
+  const auto value_id = attribute_vector()->get(chunk_offset);
+  if (value_id == null_value_id()) {
+    return std::nullopt;
+  }
+  return value_of_value_id(value_id);
 }
 
 template <typename T>
 const std::vector<T>& DictionarySegment<T>::dictionary() const {
-  // Implementation goes here
-  Fail("Implementation is missing.");
+  return _dictionary;
 }
 
 template <typename T>
-std::shared_ptr<const AbstractAttributeVector> DictionarySegment<T>::attribute_vector() const {
-  // Implementation goes here
-  Fail("Implementation is missing.");
+std::shared_ptr<AbstractAttributeVector> DictionarySegment<T>::attribute_vector() const {
+  return _attribute_vector;
 }
 
 template <typename T>
 ValueID DictionarySegment<T>::null_value_id() const {
-  // Implementation goes here
-  Fail("Implementation is missing.");
+  return _segment_nullable ? static_cast<ValueID>(0) : INVALID_VALUE_ID;
 }
 
 template <typename T>
 const T DictionarySegment<T>::value_of_value_id(const ValueID value_id) const {
-  // Implementation goes here
-  Fail("Implementation is missing.");
+  Assert(!_segment_nullable || value_id != null_value_id(), "Can't retrieve value for null value.");
```

Comment:
```suggestion
  Assert(!(_segment_nullable && value_id == null_value_id()), "Can't retrieve value for null value.");
```
I feel like this might be more readable/ better depict what we are trying to detect